### DISCUSSION TITLE
How are COM classes (generated as structs) meant to be used?

### DISCUSSION THREAD

**Discussion by jnm2** [microsoft/CsWin32#323] [Q&A]
Using 0.1.506-beta, I couldn't figure out how to use the generated `ShellLink` struct other than by stringing together `Activator.CreateInstance`, `Type.GetTypeFromCLSID`, and `typeof(ShellLink).GUID`. Is this the intended usage? Is nicer syntax thinkable?

##### NativeMethods.txt
```
IPersistFile
IShellLinkW
ShellLink
```

##### Utils.cs
```cs
using System;
using System.IO;
using Windows.Win32;
using Windows.Win32.System.Com;
using Windows.Win32.UI.Shell;

public static class Utils
{
    public static void CreateShortcut(string shortcutPath, string targetPath)
    {
        var shellLink = (IShellLinkW)Activator.CreateInstance(Type.GetTypeFromCLSID(typeof(ShellLink).GUID, throwOnError: true)!)!;

        shellLink.SetPath(targetPath);
        shellLink.SetWorkingDirectory(Path.GetDirectoryName(targetPath)!);

        ((IPersistFile)shellLink).Save(shortcutPath, fRemember: false);
    }
}
```

ShellLink.g.cs contains:

```cs
[Guid("00021401-0000-0000-C000-000000000046")]internal partial struct ShellLink
{
}
```

**Comment by jnm2** [comment]
I was able to achieve this syntax and it ran successfully, so this is at least an option to consider:

```cs
public static void CreateShortcut(string shortcutPath, string targetPath)
{
    var shellLink = new IShellLinkW();

    shellLink.SetPath(targetPath);
    shellLink.SetWorkingDirectory(Path.GetDirectoryName(targetPath)!);

    ((IPersistFile)shellLink).Save(shortcutPath, fRemember: false);
}
```

To make this work, all I had to do was apply `[CoClass(typeof(ShellLink))]` to `IShellLinkW` and change the declaration of ShellLink to this:

```cs
[ComImport, Guid("00021401-0000-0000-C000-000000000046")]
// Only one or the other of these following attributes seems to be needed.
[ClassInterface(ClassInterfaceType.None)]
[TypeLibType(TypeLibTypeFlags.FCanCreate)]
internal class ShellLink { }
```

The `new IShellLinkW()` syntax requires having something present in the metadata relating IShellLinkW to ShellLink. If that's not possible, or if the syntax below is more desirable, the same ShellLink definition above can be used *without* adding a `CoClass` attribute to IShellLinkW to enable this syntax which I also ran successfully:

```cs
public static void CreateShortcut(string shortcutPath, string targetPath)
{
    var shellLink = (IShellLinkW)new ShellLink();

    shellLink.SetPath(targetPath);
    shellLink.SetWorkingDirectory(Path.GetDirectoryName(targetPath)!);

    ((IPersistFile)shellLink).Save(shortcutPath, fRemember: false);
}
```

**Reply by jnm2** [reply]
@AArnott I'd be willing to contribute this fix. Should I convert this discussion to an issue?

**Reply by AArnott** [reply]
We have talked internally about how to make such scenarios easier. The method you suggest aligns with one of the options we discussed, I think. But this isn't an area I have a lot of expertise in yet, so I'd like to hold off till I get some ratifying votes from MS folks that are more familiar with this area.

Yes, go ahead and convert to (or file) an issue so we can track better.

**Reply by jnm2** [reply]
I forgot to create an issue, but one just showed up: https://github.com/microsoft/CsWin32/issues/453
